src/sort.rs

```diff
@@ -143,42 +146,73 @@ where
         get_many_from_sorted_mut_unchecked(self, &deduped_indexes)
     }
 
-    fn partition_mut(&mut self, pivot_index: usize) -> usize
+    fn partition_mut(&mut self) -> (usize, usize)
     where
         A: Ord + Clone,
         S: DataMut,
     {
-        let pivot_value = self[pivot_index].clone();
-        self.swap(pivot_index, 0);
-        let n = self.len();
-        let mut i = 1;
-        let mut j = n - 1;
-        loop {
-            loop {
-                if i > j {
-                    break;
-                }
-                if self[i] >= pivot_value {
-                    break;
-                }
-                i += 1;
+        let lowermost = 0;
+        let uppermost = self.len() - 1;
+        if self.len() < 42 {
+            // Sort outermost elements and use them as pivots.
+            if self[lowermost] > self[uppermost] {
+                self.swap(lowermost, uppermost);
             }
-            while pivot_value <= self[j] {
-                if j == 1 {
-                    break;
+        } else {
+            // Sample indexes of 5 evenly spaced elements around the center element.
+            let mut samples = [0; 5];
+            // Assume array of at least 7 elements.
+            let seventh = self.len() / (samples.len() + 2);
+            samples[2] = self.len() / 2;
+            samples[1] = samples[2] - seventh;
+            samples[0] = samples[1] - seventh;
+            samples[3] = samples[2] + seventh;
+            samples[4] = samples[3] + seventh;
+            // Use insertion sort for sample elements by looking up their indexes.
+            for mut index in 1..samples.len() {
+                while index > 0 && self[samples[index - 1]] > self[samples[index]] {
+                    self.swap(samples[index - 1], samples[index]);
+                    index -= 1;
                 }
-                j -= 1;
             }
-            if i >= j {
-                break;
-            } else {
-                self.swap(i, j);
-                i += 1;
-                j -= 1;
+            // Use 1st and 3rd element of sorted sample as skewed pivots.
```

Comment:
This seems like a good strategy. It's possible to design a pathological input for this strategy, but it would look pretty weird and seems unlikely to occur accidentally. Is there a reference for this strategy, or did you come up with it?

Reply:
A symmetric strategy is used by the JDK 7 implementation, they use the 2nd and 4th of five. This [master thesis](https://www.wild-inter.net/publications/wild-2012-print.pdf) suggests at page 183 in the second paragraph starting with "Finally, the case k = 5", where k is the sample length, that the 1st and 3rd of five is a better choice. But in the end it depends on the input distribution.